src/test/java/com/commercetools/sync/sdk2/taxcategories/helpers/TaxCategoryBatchValidatorTest.java

```diff
@@ -0,0 +1,197 @@
+import com.commercetools.sync.sdk2.taxcategories.TaxCategorySyncOptionsBuilder;
+import com.neovisionaries.i18n.CountryCode;
+import java.util.*;
+import javax.annotation.Nonnull;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class TaxCategoryBatchValidatorTest {
+  private List<String> errorCallBackMessages;
+  private TaxCategorySyncOptions syncOptions;
+  private TaxCategorySyncStatistics syncStatistics;
+
+  @BeforeEach
+  void setup() {
+    errorCallBackMessages = new ArrayList<>();
+    final ProjectApiRoot ctpClient = mock(ProjectApiRoot.class);
+    syncOptions =
+        TaxCategorySyncOptionsBuilder.of(ctpClient)
+            .errorCallback(
+                (exception, oldResource, newResource, actions) -> {
+                  errorCallBackMessages.add(exception.getMessage());
+                })
+            .build();
+    syncStatistics = new TaxCategorySyncStatistics();
+  }
+
+  @Test
+  void validateAndCollectReferencedKeys_WithEmptyDraft_ShouldHaveEmptyResult() {
+    final Set<TaxCategoryDraft> validDrafts = getValidDrafts(emptyList());
+
+    assertThat(validDrafts).isEmpty();
+    assertThat(errorCallBackMessages).isEmpty();
+  }
+
+  @Test
+  void
+      validateAndCollectReferencedKeys_WithNullTaxCategoryDraft_ShouldHaveValidationErrorAndEmptyResult() {
+    final Set<TaxCategoryDraft> validDrafts = getValidDrafts(Collections.singletonList(null));
+
+    assertThat(errorCallBackMessages).hasSize(1);
+    assertThat(errorCallBackMessages.get(0)).isEqualTo(TAX_CATEGORY_DRAFT_IS_NULL);
+    assertThat(validDrafts).isEmpty();
+  }
+
+  @Test
+  void
+      validateAndCollectReferencedKeys_WithTaxCategoryDraftWithNullKey_ShouldHaveValidationErrorAndEmptyResult() {
+    final TaxCategoryDraft taxCategoryDraft = mock(TaxCategoryDraft.class);
+    final Set<TaxCategoryDraft> validDrafts =
+        getValidDrafts(Collections.singletonList(taxCategoryDraft));
+
+    assertThat(errorCallBackMessages).hasSize(1);
+    assertThat(errorCallBackMessages.get(0))
+        .isEqualTo(format(TAX_CATEGORY_DRAFT_KEY_NOT_SET, taxCategoryDraft.getName()));
+    assertThat(validDrafts).isEmpty();
+  }
+
+  @Test
+  void
+      validateAndCollectReferencedKeys_WithTaxCategoryDraftWithEmptyKey_ShouldHaveValidationErrorAndEmptyResult() {
+    final TaxCategoryDraft taxCategoryDraft = mock(TaxCategoryDraft.class);
+    when(taxCategoryDraft.getKey()).thenReturn(EMPTY);
+    final Set<TaxCategoryDraft> validDrafts =
+        getValidDrafts(Collections.singletonList(taxCategoryDraft));
+
+    assertThat(errorCallBackMessages).hasSize(1);
+    assertThat(errorCallBackMessages.get(0))
+        .isEqualTo(format(TAX_CATEGORY_DRAFT_KEY_NOT_SET, taxCategoryDraft.getName()));
+    assertThat(validDrafts).isEmpty();
+  }
+
+  @Test
+  void validateAndCollectReferencedKeys_WithMixOfValidAndInvalidDrafts_ShouldValidateCorrectly() {
```

Comment:
Please add validation tests for:
1. different country, same state
2. country is null

Reply:
[52ea8d5](https://github.com/commercetools/commercetools-sync-java/pull/1039/commits/52ea8d5f84c71032f5d7e6fdf7957456a8222c1f)